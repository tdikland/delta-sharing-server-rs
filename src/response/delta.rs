use std::io::Write;

use axum::response::{IntoResponse, Response};
use bytes::{BufMut, BytesMut};
use http::{header, StatusCode};
use serde::Serialize;

use crate::{
    reader::TableMetadata,
    signer::{SignedDataFile, SignedTableData},
};

use deltalake::kernel::{
    Add as DeltaAdd, AddCDCFile as DeltaCdf, Metadata as DeltaMetadata, Protocol as DeltaProtocol,
    Remove as DeltaRemove,
};

pub struct DeltaResponse {
    version: u64,
    protocol: DeltaResponseLine,
    metadata: DeltaResponseLine,
    lines: Vec<DeltaResponseLine>,
}

impl IntoResponse for DeltaResponse {
    fn into_response(self) -> Response {
        let mut buf = BytesMut::new().writer();

        serde_json::to_writer(&mut buf, &self.protocol).unwrap();
        buf.write_all(b"\n").unwrap();

        serde_json::to_writer(&mut buf, &self.metadata).unwrap();
        buf.write_all(b"\n").unwrap();

        for line in self.lines {
            serde_json::to_writer(&mut buf, &line).unwrap();
            buf.write_all(b"\n").unwrap();
        }

        let version = self.version.to_string();
        let headers = [
            (
                header::CONTENT_TYPE.as_str(),
                "application/x-ndjson; charset=utf-8",
            ),
            ("Delta-Table-Version", version.as_ref()),
        ];

        (StatusCode::OK, headers, buf.into_inner()).into_response()
    }
}

impl From<TableMetadata> for DeltaResponse {
    fn from(table_metadata: TableMetadata) -> Self {
        Self {
            version: table_metadata.version,
            protocol: DeltaResponseLine::Protocol(table_metadata.protocol.into()),
            metadata: DeltaResponseLine::Metadata(table_metadata.metadata.into()),
            lines: vec![],
        }
    }
}

impl From<SignedTableData> for DeltaResponse {
    fn from(signed_table_data: SignedTableData) -> Self {
        let lines = signed_table_data.data.into_iter().map(Into::into).collect();

        Self {
            version: signed_table_data.version,
            protocol: DeltaResponseLine::Protocol(signed_table_data.protocol.into()),
            metadata: DeltaResponseLine::Metadata(signed_table_data.metadata.into()),
            lines,
        }
    }
}

#[derive(Debug, Clone, Serialize)]
enum DeltaResponseLine {
    Protocol(ProtocolResponseLine),
    Metadata(MetadataResponseLine),
    File(FileResponseLine),
}

impl From<SignedDataFile> for DeltaResponseLine {
    fn from(signed_data_file: SignedDataFile) -> Self {
        match signed_data_file {
            SignedDataFile::File(file) => DeltaResponseLine::File(file.into()),
            SignedDataFile::Add(add) => DeltaResponseLine::File(add.into()),
            SignedDataFile::Cdf(cdf) => DeltaResponseLine::File(cdf.into()),
            SignedDataFile::Remove(remove) => DeltaResponseLine::File(remove.into()),
        }
    }
}

#[derive(Debug, Clone, Serialize)]
struct ProtocolResponseLine {
    delta_protocol: DeltaProtocol,
}

impl From<DeltaProtocol> for ProtocolResponseLine {
    fn from(delta_protocol: DeltaProtocol) -> Self {
        Self { delta_protocol }
    }
}

#[derive(Debug, Clone, Serialize)]
struct MetadataResponseLine {
    delta_metadata: DeltaMetadata,
    version: Option<u64>,
    size: Option<u64>,
    num_files: Option<u64>,
}

impl From<DeltaMetadata> for MetadataResponseLine {
    fn from(delta_metadata: DeltaMetadata) -> Self {
        Self {
            delta_metadata,
            version: None,
            size: None,
            num_files: None,
        }
    }
}

#[derive(Debug, Clone, Serialize)]
enum SingleAction {
    Add(DeltaAdd),
    Cdf(DeltaCdf),
    Remove(DeltaRemove),
}

#[derive(Debug, Clone, Serialize)]
struct FileResponseLine {
    id: String,
    deletion_vector_file_id: Option<String>,
    version: Option<u64>,
    timestamp: Option<u64>,
    expiration_timestamp: Option<u64>,
    delta_single_action: SingleAction,
}

impl From<DeltaAdd> for FileResponseLine {
    fn from(add: DeltaAdd) -> Self {
        Self {
            id: String::from("TODO"),
            deletion_vector_file_id: None,
            version: None,
            timestamp: None,
            expiration_timestamp: None,
            delta_single_action: SingleAction::Add(add),
        }
    }
}

impl From<DeltaCdf> for FileResponseLine {
    fn from(add_cdc_file: DeltaCdf) -> Self {
        Self {
            id: String::from("TODO"),
            deletion_vector_file_id: None,
            version: None,
            timestamp: None,
            expiration_timestamp: None,
            delta_single_action: SingleAction::Cdf(add_cdc_file),
        }
    }
}

impl From<DeltaRemove> for FileResponseLine {
    fn from(remove: DeltaRemove) -> Self {
        Self {
            id: String::from("TODO"),
            deletion_vector_file_id: None,
            version: None,
            timestamp: None,
            expiration_timestamp: None,
            delta_single_action: SingleAction::Remove(remove),
        }
    }
}
