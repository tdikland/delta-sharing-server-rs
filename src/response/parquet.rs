use axum::response::{IntoResponse, Response};
use bytes::{BufMut, BytesMut};
use delta_kernel::actions::{Add, Format, Metadata, Protocol};
use http::{header, StatusCode};
use serde::Serialize;
use std::{collections::HashMap, io::Write, sync::Arc};

use crate::{
    reader::{TableData, TableMeta},
    signer::UrlSigner,
};

#[derive(Debug, Clone, Serialize)]
pub struct ParquetResponse {
    version: u64,
    protocol: ParquetResponseLine,
    metadata: ParquetResponseLine,
    lines: Vec<ParquetResponseLine>,
}

impl ParquetResponse {
    pub async fn sign(mut self, table_root: &str, signer: Arc<dyn UrlSigner>) -> Self {
        for line in self.lines.iter_mut() {
            line.sign(table_root, signer.clone()).await;
        }
        self
    }
}

impl IntoResponse for ParquetResponse {
    fn into_response(self) -> Response {
        let mut buf = BytesMut::new().writer();

        serde_json::to_writer(&mut buf, &self.protocol).unwrap();
        buf.write_all(b"\n").unwrap();

        serde_json::to_writer(&mut buf, &self.metadata).unwrap();
        buf.write_all(b"\n").unwrap();

        for line in self.lines {
            serde_json::to_writer(&mut buf, &line).unwrap();
            buf.write_all(b"\n").unwrap();
        }

        let version = self.version.to_string();
        let headers = [
            (
                header::CONTENT_TYPE.as_str(),
                "application/x-ndjson; charset=utf-8",
            ),
            ("Delta-Table-Version", version.as_ref()),
        ];

        (StatusCode::OK, headers, buf.into_inner()).into_response()
    }
}

impl From<TableMeta> for ParquetResponse {
    fn from(meta: TableMeta) -> Self {
        let protocol = ProtocolLine::from_protocol(meta.protocol().clone().into_inner());
        let metadata = MetadataLine::from_metadata_with_opts(
            meta.metadata().clone().into_inner(),
            None,
            None,
            None,
        );

        Self {
            version: meta.version(),
            protocol: protocol.into(),
            metadata: metadata.into(),
            lines: vec![],
        }
    }
}

impl From<TableData> for ParquetResponse {
    fn from(data: TableData) -> Self {
        let protocol = ProtocolLine::from_protocol(data.protocol().clone().into_inner());
        let metadata = MetadataLine::from_metadata_with_opts(
            data.metadata().clone().into_inner(),
            None,
            None,
            None,
        );

        let mut lines = vec![];
        for file in data.data().into_iter().cloned() {
            lines.push(FileLine::from_add_with_opts(file.into_inner(), None, None, None).into());
        }
        Self {
            version: data.version(),
            protocol: protocol.into(),
            metadata: metadata.into(),
            lines,
        }
    }
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
enum ParquetResponseLine {
    Protocol(ProtocolLine),
    #[serde(rename = "metaData")]
    Metadata(MetadataLine),
    File(FileLine),
}

impl ParquetResponseLine {
    async fn sign(&mut self, table_root: &str, signer: Arc<dyn UrlSigner>) {
        match self {
            Self::File(file) => {
                file.sign(table_root, signer).await;
            }
            _ => {}
        }
    }
}

impl From<ProtocolLine> for ParquetResponseLine {
    fn from(protocol: ProtocolLine) -> Self {
        Self::Protocol(protocol)
    }
}

impl From<MetadataLine> for ParquetResponseLine {
    fn from(metadata: MetadataLine) -> Self {
        Self::Metadata(metadata)
    }
}

impl From<FileLine> for ParquetResponseLine {
    fn from(file: FileLine) -> Self {
        Self::File(file)
    }
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
struct ProtocolLine {
    min_reader_version: u32,
}

impl ProtocolLine {
    fn from_protocol(protocol: Protocol) -> Self {
        let min_reader_version = u32::try_from(protocol.min_reader_version)
            .expect("protocol min_reader_version is non-negative");

        Self { min_reader_version }
    }
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
struct ParquetResponseFormat {
    provider: String,
}

impl From<Format> for ParquetResponseFormat {
    fn from(format: Format) -> Self {
        Self {
            provider: format.provider,
        }
    }
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
struct MetadataLine {
    id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    description: Option<String>,
    format: ParquetResponseFormat,
    schema_string: String,
    partition_columns: Vec<String>,
    #[serde(skip_serializing_if = "HashMap::is_empty")]
    configuration: HashMap<String, Option<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    version: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    size: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    num_files: Option<u64>,
}

impl MetadataLine {
    fn from_metadata_with_opts(
        metadata: Metadata,
        version: Option<u64>,
        size: Option<u64>,
        num_files: Option<u64>,
    ) -> Self {
        Self {
            id: metadata.id,
            name: metadata.name,
            description: metadata.description,
            format: metadata.format.into(),
            schema_string: metadata.schema_string,
            partition_columns: metadata.partition_columns,
            configuration: metadata.configuration,
            version,
            size,
            num_files,
        }
    }
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
struct FileLine {
    url: String,
    id: String,
    partition_values: HashMap<String, Option<String>>,
    size: u64,
    #[serde(skip_serializing_if = "Option::is_none")]
    stats: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    version: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    timestamp: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    expiration_timestamp: Option<i64>,
}

impl FileLine {
    fn from_add_with_opts(
        add: Add,
        version: Option<u64>,
        timestamp: Option<u64>,
        expiration_timestamp: Option<i64>,
    ) -> Self {
        let file_id = format!("{:x}", md5::compute(add.path.as_bytes()));
        let size = u64::try_from(add.size).expect("file size is non-negative");
        Self {
            url: add.path,
            id: file_id,
            partition_values: add.partition_values,
            size,
            stats: add.stats,
            version,
            timestamp,
            expiration_timestamp,
        }
    }

    async fn sign(&mut self, root_path: &str, signer: Arc<dyn UrlSigner>) {
        let file_url = format!("{}/{}", root_path, self.url);
        let signed_url = signer.sign_url(&file_url).await;

        self.url = signed_url.url().to_string();
        self.expiration_timestamp = Some(signed_url.expires_at().timestamp_millis());
    }
}

#[cfg(test)]
mod test {
    use std::time::Duration;

    use chrono::{TimeZone, Utc};
    use delta_kernel::actions::Format;
    use insta::assert_json_snapshot;
    use mockall::predicate::eq;

    use crate::signer::{MockUrlSigner, SignedUrl};

    use super::*;

    #[test]
    fn build_protocol_parquet_line() {
        let protocol = Protocol {
            min_reader_version: 1,
            min_writer_version: 2,
            reader_features: None,
            writer_features: None,
        };

        let protocol_line = ProtocolLine::from_protocol(protocol);
        let line = ParquetResponseLine::Protocol(protocol_line);

        assert_json_snapshot!(line);
    }

    #[test]
    fn build_metadata_parquet_line() {
        let metadata = Metadata {
            id: String::from("f8d5c169-3d01-4ca3-ad9e-7dc3355aedb2"),
            name: Some(String::from("name")),
            description: Some(String::from("description")),
            format: Format {
                provider: String::from("parquet"),
                options: HashMap::new(),
            },
            schema_string: String::from("{\"type\":\"struct\",\"fields\":[{\"name\":\"eventTime\",\"type\":\"timestamp\",\"nullable\":true,\"metadata\":{}}]}"),
            partition_columns: vec![String::from("date")],
            created_time: Some(1619824428000),
            configuration: HashMap::from_iter([("opt1".to_owned(), Some("true".to_owned()))]),
        };

        let mline = MetadataLine::from_metadata_with_opts(metadata, Some(1), Some(123456), Some(5));
        let line = ParquetResponseLine::Metadata(mline);

        assert_json_snapshot!(line);
    }

    #[test]
    fn build_file_parquet_line() {
        let add = Add {
            path: "key.snappy.parquet".to_owned(),
            partition_values: HashMap::from_iter([(
                "date".to_owned(),
                Some("2021-04-28".to_owned()),
            )]),
            size: 573,
            stats: Some("{\"numRecords\":1}".to_owned()),
            modification_time: 1619824428000,
            data_change: false,
            tags: HashMap::new(),
            deletion_vector: None,
            base_row_id: None,
            default_row_commit_version: None,
            clustering_provider: None,
        };

        let file_line =
            FileLine::from_add_with_opts(add, Some(1), Some(1619824428000), Some(1619824430000));
        let line = ParquetResponseLine::File(file_line);
        assert_json_snapshot!(line);
    }

    #[tokio::test]
    async fn sign_file_response_line() {
        let add = Add {
            path: "key.snappy.parquet".to_owned(),
            partition_values: HashMap::new(),
            size: 573,
            stats: None,
            modification_time: 1619824428000,
            data_change: false,
            tags: HashMap::new(),
            deletion_vector: None,
            base_row_id: None,
            default_row_commit_version: None,
            clustering_provider: None,
        };
        let mut file_line = FileLine::from_add_with_opts(add, None, None, None);

        let mut mock_signer = MockUrlSigner::new();
        mock_signer
            .expect_sign_url()
            .with(eq("s3://bucket/prefix/key.snappy.parquet"))
            .once()
            .returning(|_| {
                let url = String::from("s3://bucket/prefix/key.snappy.parquet?sig=foo_signature");
                let valid_from = Utc.timestamp_opt(1_610_000_000, 0).unwrap();
                let expiration = Duration::from_secs(3600);
                SignedUrl::new(url, valid_from, expiration)
            });

        file_line
            .sign("s3://bucket/prefix", Arc::new(mock_signer))
            .await;

        assert_eq!(
            file_line.url,
            "s3://bucket/prefix/key.snappy.parquet?sig=foo_signature"
        );
        assert_eq!(file_line.expiration_timestamp, Some(1_610_003_600_000));
        assert_json_snapshot!(file_line);
    }
}
