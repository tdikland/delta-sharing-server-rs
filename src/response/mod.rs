use std::sync::Arc;

use axum::http::HeaderName;
use axum::response::Response;
use axum::{
    http::{header, StatusCode},
    response::IntoResponse,
    Json,
};
use serde::Serialize;

use self::delta::DeltaResponse;
use self::parquet::ParquetResponse;
use crate::catalog::{Page, Schema as SchemaInfo, Share as ShareInfo, Table as TableInfo};
use crate::reader::TableVersionNumber;
use crate::signer::UrlSigner;

pub mod delta;
pub mod parquet;

static DELTA_TABLE_VERSION: HeaderName = HeaderName::from_static("delta-table-version");

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Share {
    name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    id: Option<String>,
}

impl From<ShareInfo> for Share {
    fn from(value: ShareInfo) -> Self {
        Self {
            name: value.name().to_owned(),
            id: value.id().map(|id| id.to_owned()),
        }
    }
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ListSharesResponse {
    items: Vec<Share>,
    #[serde(skip_serializing_if = "Option::is_none")]
    next_page_token: Option<String>,
}

impl From<Page<ShareInfo>> for ListSharesResponse {
    fn from(value: Page<ShareInfo>) -> Self {
        let (items, next_page_token) = value.into_parts();

        Self {
            items: items.into_iter().map(Share::from).collect(),
            next_page_token,
        }
    }
}

impl IntoResponse for ListSharesResponse {
    fn into_response(self) -> Response {
        (
            StatusCode::OK,
            [(header::CONTENT_TYPE, "application/json; charset=utf-8")],
            Json(self),
        )
            .into_response()
    }
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct GetShareResponse {
    share: Share,
}

impl From<ShareInfo> for GetShareResponse {
    fn from(value: ShareInfo) -> Self {
        Self {
            share: value.into(),
        }
    }
}

impl IntoResponse for GetShareResponse {
    fn into_response(self) -> Response {
        (
            StatusCode::OK,
            [(header::CONTENT_TYPE, "application/json; charset=utf-8")],
            Json(self),
        )
            .into_response()
    }
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Schema {
    name: String,
    share: String,
}

impl From<SchemaInfo> for Schema {
    fn from(value: SchemaInfo) -> Self {
        Self {
            name: value.name().to_owned(),
            share: value.share_name().to_owned(),
        }
    }
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ListSchemasResponse {
    items: Vec<Schema>,
    #[serde(skip_serializing_if = "Option::is_none")]
    next_page_token: Option<String>,
}

impl From<Page<SchemaInfo>> for ListSchemasResponse {
    fn from(value: Page<SchemaInfo>) -> Self {
        let (items, next_page_token) = value.into_parts();

        Self {
            items: items.into_iter().map(Schema::from).collect(),
            next_page_token,
        }
    }
}

impl IntoResponse for ListSchemasResponse {
    fn into_response(self) -> Response {
        (
            StatusCode::OK,
            [(header::CONTENT_TYPE, "application/json; charset=utf-8")],
            Json(self),
        )
            .into_response()
    }
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Table {
    name: String,
    schema: String,
    share: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    share_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    id: Option<String>,
}

impl From<TableInfo> for Table {
    fn from(value: TableInfo) -> Self {
        Self {
            name: value.name().to_owned(),
            schema: value.schema_name().to_owned(),
            share: value.share_name().to_owned(),
            share_id: value.share_id().map(ToOwned::to_owned),
            id: value.id().to_owned().map(ToOwned::to_owned),
        }
    }
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ListTablesResponse {
    items: Vec<Table>,
    #[serde(skip_serializing_if = "Option::is_none")]
    next_page_token: Option<String>,
}

impl From<Page<TableInfo>> for ListTablesResponse {
    fn from(value: Page<TableInfo>) -> Self {
        let (items, next_page_token) = value.into_parts();

        Self {
            items: items.into_iter().map(Table::from).collect(),
            next_page_token,
        }
    }
}

impl IntoResponse for ListTablesResponse {
    fn into_response(self) -> Response {
        (
            StatusCode::OK,
            [(header::CONTENT_TYPE, "application/json; charset=utf-8")],
            Json(self),
        )
            .into_response()
    }
}

#[derive(Debug, Clone, Serialize)]
pub struct TableVersionResponse {
    version: TableVersionNumber,
}

impl From<TableVersionNumber> for TableVersionResponse {
    fn from(version: TableVersionNumber) -> Self {
        Self { version }
    }
}

impl IntoResponse for TableVersionResponse {
    fn into_response(self) -> Response {
        (
            StatusCode::OK,
            [(DELTA_TABLE_VERSION.clone(), self.version.version())],
        )
            .into_response()
    }
}

#[derive(Debug, Clone)]
pub enum TableActionsResponse {
    Parquet(ParquetResponse),
    Delta(DeltaResponse),
}

impl TableActionsResponse {
    pub fn new_parquet<R: Into<ParquetResponse>>(response: R) -> Self {
        Self::Parquet(response.into())
    }

    pub fn new_delta<R: Into<DeltaResponse>>(response: R) -> Self {
        Self::Delta(response.into())
    }

    pub async fn sign(self, table_root: &str, signer: Arc<dyn UrlSigner>) -> Self {
        match self {
            TableActionsResponse::Parquet(response) => {
                let signed_response = response.sign(table_root, signer).await;
                TableActionsResponse::Parquet(signed_response)
            }
            TableActionsResponse::Delta(mut response) => {
                response.sign(table_root, signer).await;
                TableActionsResponse::Delta(response)
            }
        }
    }
}

impl IntoResponse for TableActionsResponse {
    fn into_response(self) -> Response {
        match self {
            TableActionsResponse::Parquet(response) => response.into_response(),
            TableActionsResponse::Delta(response) => response.into_response(),
        }
    }
}

#[cfg(test)]
mod test {
    use axum::{
        body::to_bytes,
        http::{header::CONTENT_TYPE, HeaderValue},
    };
    use bytes::Bytes;

    use super::*;

    #[tokio::test]
    async fn list_shares_response() {
        let share_info = Page::new(
            vec![ShareInfo::builder()
                .name("share_name")
                .id("share_id")
                .build()
                .unwrap()],
            Some("page_token".to_owned()),
        );
        let response = ListSharesResponse::from(share_info);
        let res = response.into_response();

        assert_eq!(res.status(), StatusCode::OK);
        assert_eq!(
            res.headers().get(CONTENT_TYPE).unwrap(),
            HeaderValue::from_static("application/json; charset=utf-8")
        );
        assert_eq!(
            to_bytes(res.into_body(), 1000).await.unwrap(),
            Bytes::from(
                r#"{"items":[{"name":"share_name","id":"share_id"}],"nextPageToken":"page_token"}"#
            )
        );
    }

    #[tokio::test]
    async fn get_share_response() {
        let share_info = ShareInfo::builder()
            .name("share_name")
            .id("share_id")
            .build()
            .unwrap();
        let response = GetShareResponse::from(share_info);
        let res = response.into_response();

        assert_eq!(res.status(), StatusCode::OK);
        assert_eq!(
            res.headers().get(CONTENT_TYPE).unwrap(),
            HeaderValue::from_static("application/json; charset=utf-8")
        );
        assert_eq!(
            to_bytes(res.into_body(), 1000).await.unwrap(),
            Bytes::from(r#"{"share":{"name":"share_name","id":"share_id"}}"#)
        );
    }

    #[tokio::test]
    async fn list_schemas_response() {
        let schema_info = Page::new(
            vec![SchemaInfo::builder()
                .name("schema_name")
                .share_name("share_name")
                .build()
                .unwrap()],
            Some("page_token".to_owned()),
        );
        let response = ListSchemasResponse::from(schema_info);
        let res = response.into_response();

        assert_eq!(res.status(), StatusCode::OK);
        assert_eq!(
            res.headers().get(CONTENT_TYPE).unwrap(),
            HeaderValue::from_static("application/json; charset=utf-8")
        );
        assert_eq!(
            to_bytes(res.into_body(), 1000).await.unwrap(),
            Bytes::from(
                r#"{"items":[{"name":"schema_name","share":"share_name"}],"nextPageToken":"page_token"}"#
            )
        );
    }

    #[tokio::test]
    async fn list_tables_response() {
        let table_info = Page::new(
            vec![TableInfo::builder()
                .name("table_name")
                .schema_name("schema_name")
                .share_name("share_name")
                .id("table_id")
                .share_id("share_id")
                .storage_path("not important here")
                .build()
                .unwrap()],
            Some("page_token".to_owned()),
        );
        let response = ListTablesResponse::from(table_info);
        let res = response.into_response();

        assert_eq!(res.status(), StatusCode::OK);
        assert_eq!(
            res.headers().get(CONTENT_TYPE).unwrap(),
            HeaderValue::from_static("application/json; charset=utf-8")
        );
        assert_eq!(
            to_bytes(res.into_body(), 1000).await.unwrap(),
            Bytes::from(
                r#"{"items":[{"name":"table_name","schema":"schema_name","share":"share_name","shareId":"share_id","id":"table_id"}],"nextPageToken":"page_token"}"#
            )
        );
    }

    #[tokio::test]
    async fn table_version_response() {
        let response = TableVersionResponse {
            version: TableVersionNumber::new(123),
        };
        let res = response.into_response();

        assert_eq!(res.status(), StatusCode::OK);
        assert_eq!(
            res.headers().get("Delta-Table-Version").unwrap(),
            HeaderValue::from_static("123")
        );
    }
}
