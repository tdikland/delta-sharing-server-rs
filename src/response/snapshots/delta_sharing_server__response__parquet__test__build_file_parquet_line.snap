---
source: src/response/parquet.rs
expression: line
---
{
  "file": {
    "url": "key.snappy.parquet",
    "id": "0d166c1994fcdd317b5f18e2cfab52a0",
    "partitionValues": {
      "date": "2021-04-28"
    },
    "size": 573,
    "stats": "{\"numRecords\":1}",
    "version": 1,
    "timestamp": 1619824428000,
    "expirationTimestamp": 1619824430000
  }
}
